### PR TITLE
Use "move" for IEMGUIs / Text Objects / Data Structures' Scalars and graphs instead of redrawing or using 'coords'

### PR DESCRIPTION
**Pull request by porres**
For now, only text objects, which partially closes https://github.com/pure-data/pure-data/issues/1628

more to come...

EDIT: iemguis and graphs taken care of already, only Data Structures left to go.

### CODE REVIEW

File: src/g_text.c [hunk 1]
```diff
@@ -1251,8 +1251,10 @@ static void text_displace(t_gobj *z, t_glist *glist,
     {
         t_rtext *y = glist_findrtext(glist, x);
         rtext_displace(y, glist->gl_zoom * dx, glist->gl_zoom * dy);
-        text_drawborder(x, glist, rtext_gettag(y),
-            rtext_width(y), rtext_height(y), 0);
+        sys_vgui(".x%lx.c move %sR %d %d\n", glist_getcanvas(glist), rtext_gettag(y),
+                 dx * glist->gl_zoom, dy * glist->gl_zoom);
+        sys_vgui(".x%lx.c move %siolet %d %d\n", glist_getcanvas(glist), rtext_gettag(y),
+                 dx * glist->gl_zoom, dy * glist->gl_zoom);
```
Comment (sebshader) [inline src/g_text.c:1257]: I think there might be a way to do this w/o creating a new iolet tag.. i'll think about this

Comment (porres) [reply src/g_text.c:1257]: I though about adding a global tag as I did with many others, but it got complicated and I gave up thinking it couldn't be done sanely. Please have a go at it though to make sure :)

**Comment by sebshader** [comment]
too bad we can't use moveto without dropping tk 8.5

**Comment by sebshader** [comment]
@porres the canvas moveto command was only introduced in tk 8.6 so if we want to support tk 8.5 we can't use it. It would greatly simplify a lot of this. for instance for the iemgui stuff we wouldn't have to add arguments to the draw function, we could simply call moveto the correct (current) coordinates

**Comment by porres** [comment]
And what is the advantage of using "move" to "coords", and "moveto" to "move"?

I'm assuming the "move" command is more efficient but I didn't get where I think it's important yet, which would be moving objects within graphs and not redrawinf data structures, can you maybe help me with that @sebshader ?

**Comment by sebshader** [comment]
@porres yeah, more efficient and simpler to write. moveto uses absolute coordinates, and move uses relative ones.

**Comment by porres** [comment]
> I'm unsure if it's worth refactoring the iemguis 

I already did it! It wasn't bad

**Comment by sebshader** [comment]
> > I'm unsure if it's worth refactoring the iemguis
> 
> I already did it! It wasn't bad

yes, but in doing so certain code gets more involved. Not only does the draw code need 2 extra arguments for the delta values, but in using only move everywhere there is slightly more c-side code like:
```
int xpos = (int)atom_getfloatarg(0, ac, av);
     int ypos = (int)atom_getfloatarg(1, ac, av);
     int dx = xpos - iemgui->x_obj.te_xpix;
     int dy = ypos - iemgui->x_obj.te_ypix;
     iemgui->x_obj.te_xpix = xpos;
     iemgui->x_obj.te_ypix = ypos;
     if(glist_isvisible(iemgui->x_glist))
     {
         (*iemgui->x_draw)(x, iemgui->x_glist, IEM_GUI_DRAW_MODE_MOVE);
         (*iemgui->x_draw)(x, iemgui->x_glist, IEM_GUI_DRAW_MODE_DISPLACE, dx*zoom, dy*zoom);
         canvas_fixlinesfor(iemgui->x_glist, (t_text*)x);
     }
```
I think a good compromise might be to use macros for generics and set specific widgetbehaviors for each iemgui. I'm looking into that a bit now

**Comment by porres** [comment]
> I think a good compromise might be to use macros for generics and set specific widgetbehaviors for each iemgui.

so much more refactoring instead? I agree it's not ideal, but I was quite little intrusive

**Comment by umlaeute** [comment]
as always i think that iemgui's should be refactored on a BIG scale, rather than doing some micro-optimisations.

with *BIG scale* i mean: move all the drawing code to the Tcl/Tk side, and use some well defined interface (ideally: mimicking the GUI->Core interface) to control that from the Core.

**Comment by umlaeute** [comment]
re *BIG scale* refactoring:

this would then get rid of practically all the `class_setwidget` functionality: all objects (however fancy their GUI) would use the same displace/select/... functions (on the core-side), it's only the GUI side that would differentiate.

**Comment by porres** [comment]
I can gladly revert my commits for the iemguis change in favor of https://github.com/pure-data/pure-data/pull/1633 how do I do that?

**Comment by porres** [comment]
I guess I can just start a new cleaner PR instead with a single commit ;)

**Comment by porres** [comment]
closing in favor of https://github.com/pure-data/pure-data/pull/1635

**Pull request closed without merging**